src/lib/FreeC/Pass/CompletePatternPass.hs

```diff
@@ -0,0 +1,123 @@





















































+--





















































+--   should not pass the check.





















































+--





















































+--   == Example 2





















































+--   The following declaration with redundant alternavies





















































+--





















































+--   > redundant :: Just Bool -> Just Bool





















































+--   > redundant (on :: Just Bool) = case (on :: Just Bool) of





















































+--   >                  {Some a -> Some (False)





















































+--   >                  ,None -> None





















































+--   >                  ,Some b -> Some (True)}





















































+--





















































+--   should not pass the check either.





















































+--





















































+--   == Example 3





















































+--   The following declaration where the scrutinee is a function





















































+--





















































+--   > case_id = case \x -> x  of





















































+--





















































+--   should not pass the check.





















































+
+module FreeC.Pass.CompletePatternPass
+  ( completePatternPass
+  , checkPatternFuncDecl
+  )
+where
+
+import           Control.Monad                  ( unless )
+
+import           FreeC.Environment
+import           FreeC.Environment.Entry
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+
+-- | Checks that all functions of a given module have complete pattern matching.





















































+--   The pattern matching is complete if there is exactly one case alternative





















































+--   for each constructor of the corresponding type.





















































+completePatternPass :: Pass IR.Module
+completePatternPass ast = do
+  mapM_ checkPatternFuncDecl (IR.modFuncDecls ast)
+  return ast
+
+-- | Checks a funcDecl for incomplete Patterns





















































+checkPatternFuncDecl :: IR.FuncDecl -> Converter ()
+checkPatternFuncDecl funcDecl = checkPatternExpr (IR.funcDeclRhs funcDecl)
+ where
+  checkPatternExpr :: IR.Expr -> Converter ()
```

Comment:
Even though they are not visible in the generated Haddock documentation, I'd also add documentation comments to local declarations in `where` blocks.